clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -695,29 +699,63 @@ int nextTimeoutMs() {
         private List<Call> newCalls = new LinkedList<>();
 
         /**
-         * Check if the AdminClient metadata is ready.
-         * We need to know who the controller is, and have a non-empty view of the cluster.
+         * True if we should request new metadata.
+         */
+        private final AtomicBoolean shouldRequestMetadata = new AtomicBoolean(true);
+
+        /**
+         * The last version of the metadata which we've seen.
+         * Only accessed from the worker thread.
+         */
+        private int lastSeenMetadataVersion = 0;
+
+        /**
+         * The lowest version of the metadata which is valid.
+         * Only accessed from the worker thread.
+         */
+        private int lowestValidMetadataVersion = 0;
+
+        /**
+         * The exception which we got while fetching metadata, or null if there was none.
+         * Only accessed from the worker thread.
+         */
+        private ApiException metadataAuthException = null;
+
+        /**
+         * Check if the AdminClient metadata is usable.
          *
-         * @param prevMetadataVersion       The previous metadata version which wasn't usable.
-         * @return                          null if the metadata is usable; the current metadata
-         *                                  version otherwise
+         * @return                  null, if the metadata is usable.
+         *                          StaleMetadataException, if the metadata is not current.
+         *                          AuthenticationException, if there was an authentication issue
+         *                          while fetching the metadata.
          */
-        private Integer checkMetadataReady(Integer prevMetadataVersion) {
-            if (prevMetadataVersion != null) {
-                if (prevMetadataVersion == metadata.version())
-                    return prevMetadataVersion;
+        private ApiException checkMetadata() {
+            // If someone has requested a metadata update, or if the current metadata is
+            // just the bootstrap cluster, request a new metadata update.
+            if (shouldRequestMetadata.getAndSet(false) ||
+                    metadata.fetch().isBootstrapConfigured()) {
+                if (lowestValidMetadataVersion == lastSeenMetadataVersion) {
+                    lastSeenMetadataVersion = metadata.requestUpdate();
```

Comment:
The point of this if statement is that we shouldn't keep calling Metadata#requestUpdate over and over.  We should call it once, wait for an update, and then call it again after that if needed.

Reply:
I was not questioning the need of the `if`, but the reassignment to `lastSeenMetadataVersion`, which seemed redundant given how it is updated below. I may have answered my own question however. The problem is that the version in `Metadata` is updated in two additional cases: 1) on initial bootstrapping, and 2) after a periodic refresh. We handle the first case I think because of this reassignment and the fact that the two fields are initialized to 0.

I'm not sure about the second case, however. Say, for example, that `lowestValidMetadataVersion` and `lastSeenMetadataVersion` are both 5 and no update has been requested. After the metadata max age expires, we'll automatically trigger an update and bump `lastSeenMetadataVersion` to 6. Now if we request a metadata update, nothing will happen because we'll never reach equality again. We may be able to fix it by changing to an inequality, but I'm not sure. Some testing would be helpful.